### DISCUSSION TITLE
C API ability to bind list to prepared statement

### DISCUSSION THREAD

**Discussion by karlseguin** [duckdb/duckdb#7482] [Ideas]
I'm writing a Zig driver (https://github.com/karlseguin/zuckdb.zig). It's working well, but I noticed that there seems to be no way to bind a list to a prepared statement. I think this would be useful to have.

**Comment by Tishj** [comment]
It does exist in the C++ API, and should be trivial to add to the C API, the main problem is that you'd need to duplicate all the existing bind methods to add a list variant, and even then that won't support arbitrary nesting

**Reply by karlseguin** [reply]
I think just supporting simple types would be a good start and cover a lot of common cases!

**Reply by karlseguin** [reply]
This might make ZERO sense, but what about an API that looks something like:

```c
duckdb_bind_vector vector;
// creates a vector that can hold 3 of whatever the idx parameter type is
duckdb_create_bind_vector(prepared_statement, idx, 3, &vector);

duckdb_bind_vector_insert(vector, 0, value_1);
duckdb_bind_vector_insert(vector, 1, value_2);
duckdb_bind_vector_insert(vector, 2, value_3);
```

The CAPI would do some of the heavy lifting, creating a vector of the right size based on the parameter type and the length given (`3` above) and inserting the data into vector at the given index.

A simpler API could be exposed which requires the consumer to (a) figure the correct vector size and (b) write to the vector's raw_data directly. I don't think this would be awful, since all the implementations already know how to _read_ from a vector...so writing raw bytes into it wouldn't be crazy.
